teps/0047-finallytask-execution-post-timeout.md

```diff
@@ -0,0 +1,269 @@
+|                                                     |           |               |            |                        |
+|                                                     |           |               |            |                        |
+
+
+The pipelinerun timeout take precedence over the task timeout. After 10s the task fails... And the finally task does not get the chance to execute.
+
+
+For this reason, it is currently not possible to rely on Finally tasks for any of the aforementioned use cases.
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+Enable the uses cases :
+
+- Cleanup cluster resources after finishing (with success/failure) integration tests (Dogfooding Scenario)
+- Update Pull Request with what happened overall in the pipeline (pipeline level)
+- Report Test Results at the end of the test pipeline (Notifications Scenario)
+
+When a pipelinerun times out.
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Enable finally task to run when a pipeline times out.
+
+We propose to deprecate the current `timeout` as the pipelinerun level. This timeout is as of today used as the timeout of each newly created taskrun by the controller.
+
+Instead we propose to have to new flags: 
+- tasksTimeout
+- runTimeout
+
+`tasksTimeout` will define a timeout for the dag tasks. `runTimeout` will define a timeout for the whole pipelinerun. The finally tasks timeout willbe `runTimeout - tasksTimeout` with `runTimeout > tasksTimeout`.
+
+When `tasksTimeout` is not defined, `runTimeout` is used for the tasks timeout.
+
+This will enable users to manage run time behavior and make sure their finally tasks run as intended by scoping the tasks runtime period.
+
+.
+```yaml
+spec:
+  tasksTimeout: "1h0m0s"
+  runTimeout: "2h0m0s"
+  pipelineSpec:
+    tasks:
+    - name: tests
+      taskRef:
+        Name: integration-test
+    finally:
+    - name: cleanup-test
+      taskRef:
+        Name: cleanup
+```
+
+The default default-timeout-minutes configurable via configmap is kept and will be applied to the `runTimeout`.
+
+In order to keep older declaration working. We will use the [`Feature Flag for all alpha features`](https://github.com/tektoncd/community/pull/280) to enable the new feature.
+
+When we opt-in the feature, we may specify a `tasksTimeout` and a `runTimeout` flags. The will be used to define dags tasks timeout and the pipelinerun timeout. We cannot use the current `timeout` flag. It would result in a validation error.
```

Comment:
What would be the upgrade path with this approach if I have pipelines with `timeout` specified?

I think I would have to:
- delete `timeout` from existing pipelines
- enable the feature flag
- add `runTimeout` to pipelines

Possibly a different approach would be to deprecate "timeout" - if the feature flag is enabled it will be considered as a `runTimeout` instead, but if `tasksTimeout` or `runTimeout` are added, `timeout` has to be removed or it won't pass validation.

Reply:
We were thinking about the first approach. If the feature flag is enabled and you have a `timeout` specified, then you have a validation error.

So as you said, delete and replace it with `runTimeout`.